src/main/java/org/myrobotlab/service/Runtime.java

```diff
@@ -2006,11 +2005,19 @@ synchronized public static void unregister(String inName) {
 
     // FIXME - release autostarted peers ?
 
-    // last step - remove from registry
-    registry.remove(name);
+    // last step - remove from registry by making new registry
+    // thread safe way
+    Map<String, ServiceInterface> removedService = new TreeMap<>();
+    for (String key : registry.keySet()) {
+      if (!name.equals(key)) {
+        removedService.put(key, registry.get(key));
+      }
+    }
+    registry = removedService;

```

Comment:
This doesn't guarantee that all threads will immediately see the updated registry, since the CPU caches will hang on to it. Needs to be marked volatile. Additionally, any code storing the registry will only ever see the old one, never the new since it hangs onto the old reference.

Reply:
Its private, so all consuming classes outside of runtime should be using accessors.
The getRegistry really should be returning a copy, which in general should be the case of a getter on a private, otherwise what's the point, unless its immutable?  The getRegistry does have a FIXME but not including it in this pr. I'll mark it as volatile.